### DISCUSSION TITLE
twoside option with wide long tables

### DISCUSSION THREAD

**Discussion by jonhall** [lvjr/tabularray#543] [Ideas]
In a textbook I'm contributing to, there's a wide outer margin for notes. Many of my tables are longtblr and use the full pagewidth as they have many columns. 

On odd pages, the tables extend into the margin, but on even pages they do not. This is expected as the even page margin is on the right and the tables span left to right (see below and screenshot). 

even                                                                            |odd
| margin | text.................................................| binding |..................................................text | margin |

I'd like (an option for) long tables that start on an even/odd page and break onto an odd/even page to span the pagewidth.

I've been playing with redefining the \TblrNewPage as a "hook" to make this happen, but haven't been successful.

![Screenshot 2024-12-04 at 10 12 15](https://github.com/user-attachments/assets/566cc9af-4835-4cf4-b30c-57cf49bfe951)

**Comment by lvjr** [comment]
I think this is impossible: the widths of `longtblr` tables could not be changed.

**Reply by muzimuzhi** [reply]
It's not a request for variable width (of broken long table parts), but for different oversize direction in two-side documents.

**Comment by muzimuzhi** [comment]
```tex
\documentclass{article}
\usepackage[showframe, twoside, paperheight=6cm]{geometry}
\usepackage{tabularray}
\usepackage{xcolor}

\usepackage{changepage}

\ExplSyntaxOn
\cs_undefine:N \__tblr_halign_whole:Nn
\cs_new_protected:Npn \__tblr_halign_whole:Nn #1#2
  {
    \__tblr_check_odd_page:
    \use:e
      {
        \__tblr_halign_whole_aux:Nnn #1
          \__tblr_toggle_left_right:nn
            {
              \str_if_eq:nnT {#2} {l} { \hfil }
              \str_if_eq:nnT {#2} {l-hss} { \hss }
            }
            {
              \str_if_eq:nnT {#2} {r} { \hfil }
              \str_if_eq:nnT {#2} {r-hss} { \hss }
            }
      }
  }

\cs_new_protected:Npn \__tblr_halign_whole_aux:Nnn #1#2#3
  {
    \noindent
    % why not use \__tblr_prop_item:nn { inner } { width }?
    \hbox_to_wd:nn { \linewidth }
      {
        #2
        \box_use:N #1
        #3
      }
  }

\cs_new_protected:Npn \__tblr_check_odd_page: {}

% named after "/tcb/toggle left and right" key from tcolorbox
\cs_new:Npn \__tblr_toggle_left_right:nn #1#2 { {#1} {#2} }

% pity, file hooks like "package/<pkgname>/after" are not one-time hooks
\hook_gput_code:nnn { begindocument } { tblr }
  {
    \IfPackageLoadedTF {changepage}
      {
        \cs_gset_protected:Npn \__tblr_check_odd_page:
          {
            \checkoddpage
          }
        \cs_gset:Npn \__tblr_toggle_left_right:nn #1#2
          {
            \bool_lazy_all:nTF
              {
                { \l__tblr_toggle_left_right_bool }
                { \legacy_if_p:n { @twoside } } % \if@twoside
                { ! \legacy_if_p:n { oddpage } } % \ifoddpage from changepage pkg
              }
              { {#2} {#1} }
              { {#1} {#2} }
          }
      }
      {}
  }

% glue code for https://github.com/muzimuzhi/tabularray-dev
\keys_if_exist:nnTF { tblr/__outer } { halign }
  { \keys_define:nn { tblr/__outer} }
  { \keys_define:nn { tblr/outer } }
  {
    toggle-left-and-right .bool_set:N = \l__tblr_toggle_left_right_bool
  , toggle-left-and-right .default:n  = true
  }
\ExplSyntaxOff

\begin{document}
\begin{longtblr}[
  % outer option "halign" is undocumented!
  % new value "r-hss": ragged right using \hss
  halign=r-hss,
  toggle-left-and-right
]{
  hlines={blue}, vlines={blue},
  colspec={XX},
  width=\linewidth+\marginparsep+\marginparwidth
}
  a & a \\
  b & b \\
  c & c \\
  d & d \\
  e & e \\
  f & f \\
  g & g
\end{longtblr}
\end{document}
```

![image](https://github.com/user-attachments/assets/971418b5-6388-426f-a35c-d0b4fdf06f7f)

**Reply by muzimuzhi** [reply]
@jonhall BTW, how do you get the gray page frames? It's more elegant than the black ones I use, produced by `geometry` package.

**Reply by jonhall** [reply]
Many thanks, that's great. 
If I haven't understood, the page frames are from the geometry package:
%%For layout
\usepackage{geometry}
\geometry{
 showframe,
%  paperwidth=6in,
%  paperheight=9in,
%	asymmetric,
  outer=1cm,
  inner=1cm,
  top=1.5cm,
  bottom=1.5cm,
  marginparsep=8mm,
  marginparwidth=55mm,
  includemp,
  includehead,
}

**Reply by muzimuzhi** [reply]
But that draws black page frames, not the gray ones shown in your original screenshots.

**Reply by jonhall** [reply]
In that case, I don't know – sorry!

**Comment by jonhall** [comment]
That works great, thank you – see attached.

I'll be sure to acknowledge you both work in the final version of the book, if that is ok.

![Screenshot 2024-12-06 at 10 16 11](https://github.com/user-attachments/assets/487980ff-bd9d-4d9f-8c2d-eac7f96abce8)

**Reply by muzimuzhi** [reply]
Sure. My honor. 😄
